java/daikon/dcomp/BuildJDK.java

```diff
@@ -3,375 +3,500 @@
           }
-          File dir = new File(dfile, destfile.getParent());
+          File dir = new File(dest_dir, destfile.getParent());
           dir.mkdirs();
           File destpath = new File(dir, destfile.getName());
           if (verbose) System.out.println("Copying Object or non-classfile: " + destpath);
-          copyStreams(jfile.getInputStream(entry), new FileOutputStream(destpath));
+          try (InputStream in = class_stream_map.get(className)) {
+            Files.copy(in, destpath.toPath());
+          }
           continue;
         }
 
         // Get the binary for this class
-        InputStream is = jfile.getInputStream(entry);
+        InputStream is = class_stream_map.get(className);
         JavaClass jc;
         try {
-          ClassParser parser = new ClassParser(is, entryName);
+          ClassParser parser = new ClassParser(is, className);
           jc = parser.parse();
-        } catch (Exception e) {
-          throw new Error("Failed to parse entry " + entry, e);
+        } catch (Throwable e) {
+          throw new Error("Failed to parse classfile " + className, e);
         }
-        classmap.put(jc.getClassName(), jc);
-      }
-
-      if (false) {
-        processClassFile(classmap, dfile, "sun.rmi.registry.RegistryImpl_Skel");
-        System.exit(0);
-      }
 
-      // Process each file read.
-      for (String classname : classmap.keySet()) {
-        JavaClass jc = classmap.get(classname);
+        // Instrument the class file.
         try {
-          processClassFile(classmap, dfile, classname);
+          instrumentClassFile(jc, dest_dir, className);
         } catch (Throwable e) {
-          throw new Error("Couldn't instrument " + classname, e);
+          throw new Error("Couldn't instrument " + className, e);
         }
       }
 
-      // Create the DcompMarker class (used to identify instrumented calls)
+      // Convert from JDK version number to ClassFile major_version.
+      // A bit of a hack, but seems OK.
+      int java_class_version = BcelUtil.javaVersion + 44;
+      String dest_dir_name = dest_dir.getName();
+
+      // We've finished instrumenting all the class files. Now we create and
+      // the DcompMarker class which is used to identify instrumented calls.
       ClassGen dcomp_marker =
           new ClassGen(
               "java.lang.DCompMarker",
               "java.lang.Object",
-              "DCompMarker.class",
+              "daikon.dcomp.BuildJDK tool",
               Const.ACC_INTERFACE | Const.ACC_PUBLIC | Const.ACC_ABSTRACT,
               new String[0]);
+      dcomp_marker.setMinor(0);
+      dcomp_marker.setMajor(java_class_version);
       dcomp_marker
           .getJavaClass()
           .dump(
               new File(
-                  dest, "java" + File.separator + "lang" + File.separator + "DCompMarker.class"));
+                  dest_dir_name,
+                  "java" + File.separator + "lang" + File.separator + "DCompMarker.class"));
```

Comment:
I suggest abstracting this out into a new method, since it is now used 5 times in this class.  That will shorten and simplify the code.

Reply:
Sounds like a good idea.  I skipped this one for now to get everything else checked in.

Reply:
Will you make this change before we merge?  I think it will make the changes smaller and easier to read.
(No problem with doing the other changes first.)